modules/tinymce/src/themes/silver/main/ts/ui/core/ChoiceControls.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright (c) Tiny Technologies, Inc. All rights reserved.
+ * Licensed under the LGPL or a commercial license.
+ * For LGPL see License.txt in the project root for license information.
+ * For commercial licenses see https://www.tiny.cloud/
+ */
+
+import { Arr, Fun, Obj, Optional, Singleton, Type } from '@ephox/katamari';
+import { Attribute, Dimension, SugarElement, SugarNode, TransformFind } from '@ephox/sugar';
+
+import Editor from 'tinymce/core/api/Editor';
+import { ContentLanguage } from 'tinymce/core/api/SettingsTypes';
+import { Menu } from 'tinymce/core/api/ui/Ui';
+
+import * as Settings from '../../api/Settings';
+
+interface ControlSpec<T> {
+  readonly name: string;
+  readonly text: string;
+  readonly icon: string;
+
+  readonly getOptions: (editor: Editor) => Array<T>;
+  readonly normalise: (item: T) => string;
+  readonly display: (item: T) => string;
+
+  readonly getCurrent: (editor: Editor) => Optional<T>;
+  readonly setCurrent: (editor: Editor, value: T) => void;
+}
+
+const registerController = <T>(editor: Editor, spec: ControlSpec<T>) => {
```

Comment:
This is really looking the same as a bespoke select now... we should log a follow up JIRA to look at merging the two, as I believe it could be done.

Reply:
Created TINY-7642, if you want to update the description beyond my guesses for what this would entail.